File: internal/monitor/reporter.go
package monitor

import (
	"fmt"
	"io"
	"math"
	"time"

	"github.com/guptarohit/asciigraph"
	"github.com/richardcase/pinger/internal/config"
)

// DisplayMode selects the monitor's terminal output style.
type DisplayMode int

const (
	// DisplayLog is the default per-cycle text output.
	DisplayLog DisplayMode = iota
	// DisplayChart is the realtime ASCII RTT chart.
	DisplayChart
)

// ResolveDisplay validates a --display flag value against TTY availability.
// Unknown values and a non-TTY chart request both yield an actionable error.
func ResolveDisplay(mode string, isTTY bool) (DisplayMode, error) {
	switch mode {
	case "log":
		return DisplayLog, nil
	case "chart":
		if !isTTY {
			return DisplayLog, fmt.Errorf("error: --display chart requires an interactive terminal. Use --display log when piping or redirecting output.")
		}
		return DisplayChart, nil
	default:
		return DisplayLog, fmt.Errorf("error: invalid --display %q: must be \"log\" or \"chart\".", mode)
	}
}

// targetCycle is one target's data for a single probe cycle, handed to a reporter.
// Sent/Errors/AvgMs are cumulative across the run (used by logReporter); CycleMs is
// this cycle's RTT, nil on failure (used by chartReporter as a NaN gap).
type targetCycle struct {
	Label   string
	Sent    int
	Errors  int
	AvgMs   float64
	CycleMs *float64
}

// reporter renders monitor output for one probe cycle and the final summary.
type reporter interface {
	cycle(ts time.Time, rows []targetCycle)
	final(ts time.Time, rows []targetCycle)
}

// logReporter reproduces the original per-cycle text output verbatim.
type logReporter struct {
	w io.Writer
}

func (r *logReporter) cycle(ts time.Time, rows []targetCycle) {
	for _, row := range rows {
		if row.Sent == 0 {
			continue
		}
		fmt.Fprintf(r.w, "[%s] %s sent=%d errors=%d avg_rtt=%.2fms\n",
			ts.Format(time.RFC3339), row.Label, row.Sent, row.Errors, row.AvgMs)
	}
}

func (r *logReporter) final(ts time.Time, rows []targetCycle) {
	for _, row := range rows {
		fmt.Fprintf(r.w, "[%s] %s TOTAL sent=%d errors=%d avg_rtt=%.2fms\n",
			ts.Format(time.RFC3339), row.Label, row.Sent, row.Errors, row.AvgMs)
	}
	fmt.Fprintf(r.w, "[%s] monitor stopped\n", ts.Format(time.RFC3339))
}

// Terminal control sequences for the alternate screen buffer and cursor.
const (
	enterAltScreen = "\x1b[?1049h"
	leaveAltScreen = "\x1b[?1049l"
	hideCursor     = "\x1b[?25l"
	showCursor     = "\x1b[?25h"
	clearAndHome   = "\x1b[H\x1b[2J"
)

// chartPalette gives each target a visually distinct colored line. It holds ≥10
// distinct colors so all 10 max targets stay distinguishable; cycled if exceeded.
var chartPalette = []asciigraph.AnsiColor{
	asciigraph.Red, asciigraph.Green, asciigraph.Blue, asciigraph.Yellow,
	asciigraph.Cyan, asciigraph.Magenta, asciigraph.Orange, asciigraph.Purple,
	asciigraph.SpringGreen, asciigraph.DodgerBlue, asciigraph.HotPink, asciigraph.Gold,
}

// sizeFunc reports the current terminal width and height in cells.
type sizeFunc func() (width, height int)

// chartSeries is one target's rolling RTT buffer; math.NaN() marks a failed cycle.
type chartSeries struct {
	label  string
	points []float64
}

// chartReporter maintains per-target rolling buffers and redraws a combined
// asciigraph plot each cycle into the alternate screen buffer.
type chartReporter struct {
	w       io.Writer
	size    sizeFunc
	series  []*chartSeries
	index   map[string]int
	started bool
}

func newChartReporter(w io.Writer, size sizeFunc, cfg *config.Config) *chartReporter {
	cr := &chartReporter{w: w, size: size, index: make(map[string]int, len(cfg.Targets))}
	for i, t := range cfg.Targets {
		cr.series = append(cr.series, &chartSeries{label: t.Label})
		cr.index[t.Label] = i
	}
	return cr
}

func (r *chartReporter) cycle(_ time.Time, rows []targetCycle) {
	if !r.started {
		fmt.Fprint(r.w, enterAltScreen+hideCursor)
		r.started = true
	}
	width, height := r.size()
	for _, row := range rows {
		i, ok := r.index[row.Label]
		if !ok {
			continue
		}
		v := math.NaN()
		if row.CycleMs != nil {
			v = *row.CycleMs
		}
		s := r.series[i]
		s.points = append(s.points, v)
		if width > 0 && len(s.points) > width {
			s.points = s.points[len(s.points)-width:]
		}
	}
	r.render(width, height)
}

func (r *chartReporter) render(width, height int) {
	data := make([][]float64, len(r.series))
	legends := make([]string, len(r.series))
	colors := make([]asciigraph.AnsiColor, len(r.series))
	for i, s := range r.series {
		data[i] = s.points
		legends[i] = s.label
		colors[i] = chartPalette[i%len(chartPalette)]
	}

	plotWidth := width - 10
	if plotWidth < 10 {
		plotWidth = 10
	}
	plotHeight := height - 4
	if plotHeight < 3 {
		plotHeight = 3
	}

	// asciigraph derives the Y-axis from min/max; with zero finite points it would
	// divide by an infinite interval and produce a NaN row count. Skip until real data.
	if !anyFinite(data) {
		fmt.Fprint(r.w, clearAndHome)
		fmt.Fprintln(r.w, "RTT (ms) over time — waiting for data…")
		return
	}

	graph := asciigraph.PlotMany(data,
		asciigraph.Width(plotWidth),
		asciigraph.Height(plotHeight),
		asciigraph.SeriesColors(colors...),
		asciigraph.SeriesLegends(legends...),
		asciigraph.Caption("RTT (ms) over time"),
	)
	fmt.Fprint(r.w, clearAndHome)
	fmt.Fprintln(r.w, graph)
}

func anyFinite(data [][]float64) bool {
	for _, series := range data {
		for _, v := range series {
			if !math.IsNaN(v) {
				return true
			}
		}
	}
	return false
}

func (r *chartReporter) final(ts time.Time, rows []targetCycle) {
	if r.started {
		fmt.Fprint(r.w, leaveAltScreen+showCursor)
	}
	(&logReporter{w: r.w}).final(ts, rows)
}

File: src/monitor/reporter.zig
//! Monitor output reporters, mirroring internal/monitor/reporter.go: a
//! byte-for-byte log reporter and a realtime chart reporter that maintains
//! per-target rolling RTT buffers and redraws an asciigraph plot in the
//! alternate screen buffer.

const std = @import("std");
const c = std.c;
const posix = std.posix;
const rfc3339 = @import("../rfc3339.zig");
const gofmt = @import("../gofmt.zig");
const sys = @import("../sys.zig");
const asciigraph = @import("asciigraph.zig");

pub const Display = enum { log, chart };

pub const ResolveResult = union(enum) {
    ok: Display,
    not_tty,
    invalid: []const u8,
};

/// Validate a --display value against TTY availability (matches ResolveDisplay).
pub fn resolveDisplay(mode: []const u8, is_tty: bool) ResolveResult {
    if (std.mem.eql(u8, mode, "log")) return .{ .ok = .log };
    if (std.mem.eql(u8, mode, "chart")) {
        if (!is_tty) return .not_tty;
        return .{ .ok = .chart };
    }
    return .{ .invalid = mode };
}

pub fn isStdoutTty() bool {
    return c.isatty(sys.stdout_fd) != 0;
}

/// One target's data for a probe cycle. Sent/Errors/AvgMs are cumulative;
/// CycleMs is this cycle's RTT (null on failure).
pub const TargetCycle = struct {
    label: []const u8,
    sent: i64,
    errors: i64,
    avg_ms: f64,
    cycle_ms: ?f64,
};

const enter_alt_screen = "\x1b[?1049h";
const leave_alt_screen = "\x1b[?1049l";
const hide_cursor = "\x1b[?25l";
const show_cursor = "\x1b[?25h";
const clear_and_home = "\x1b[H\x1b[2J";

fn writeLogLines(gpa: std.mem.Allocator, ts: rfc3339.Time, rows: []const TargetCycle, comptime total: bool) void {
    var tb: [40]u8 = undefined;
    const tsstr = rfc3339.formatRFC3339(&tb, ts.secs);
    for (rows) |row| {
        if (!total and row.sent == 0) continue;
        var ab: [350]u8 = undefined;
        const avg = gofmt.formatFixed(&ab, row.avg_ms, 2);
        const line = if (total)
            std.fmt.allocPrint(gpa, "[{s}] {s} TOTAL sent={d} errors={d} avg_rtt={s}ms\n", .{ tsstr, row.label, row.sent, row.errors, avg }) catch return
        else
            std.fmt.allocPrint(gpa, "[{s}] {s} sent={d} errors={d} avg_rtt={s}ms\n", .{ tsstr, row.label, row.sent, row.errors, avg }) catch return;
        defer gpa.free(line);
        sys.writeAll(sys.stdout_fd, line);
    }
    if (total) {
        const stopped = std.fmt.allocPrint(gpa, "[{s}] monitor stopped\n", .{tsstr}) catch return;
        defer gpa.free(stopped);
        sys.writeAll(sys.stdout_fd, stopped);
    }
}

pub const LogReporter = struct {
    gpa: std.mem.Allocator,

    pub fn cycle(self: *LogReporter, ts: rfc3339.Time, rows: []const TargetCycle) void {
        writeLogLines(self.gpa, ts, rows, false);
    }
    pub fn final(self: *LogReporter, ts: rfc3339.Time, rows: []const TargetCycle) void {
        writeLogLines(self.gpa, ts, rows, true);
    }
};

pub const ChartReporter = struct {
    gpa: std.mem.Allocator,
    labels: [][]const u8,
    series: []std.ArrayList(f64),
    started: bool = false,

    pub fn init(gpa: std.mem.Allocator, labels: []const []const u8) !ChartReporter {
        const lbl = try gpa.alloc([]const u8, labels.len);
        const ser = try gpa.alloc(std.ArrayList(f64), labels.len);
        for (labels, 0..) |l, i| {
            lbl[i] = l;
            ser[i] = .empty;
        }
        return .{ .gpa = gpa, .labels = lbl, .series = ser };
    }

    pub fn deinit(self: *ChartReporter) void {
        for (self.series) |*s| s.deinit(self.gpa);
        self.gpa.free(self.series);
        self.gpa.free(self.labels);
    }

    pub fn cycle(self: *ChartReporter, ts: rfc3339.Time, rows: []const TargetCycle) void {
        _ = ts;
        if (!self.started) {
            sys.writeAll(sys.stdout_fd, enter_alt_screen ++ hide_cursor);
            self.started = true;
        }
        const sz = termSize();
        for (rows, 0..) |row, i| {
            if (i >= self.series.len) break;
            const v = row.cycle_ms orelse std.math.nan(f64);
            self.series[i].append(self.gpa, v) catch {};
            if (sz.w > 0 and self.series[i].items.len > sz.w) {
                const drop = self.series[i].items.len - sz.w;
                std.mem.copyForwards(f64, self.series[i].items[0..sz.w], self.series[i].items[drop..]);
                self.series[i].shrinkRetainingCapacity(sz.w);
            }
        }
        self.render(sz.w, sz.h);
    }

    fn render(self: *ChartReporter, width: usize, height: usize) void {
        _ = width;
        var plot_h: usize = if (height > 4) height - 4 else 3;
        if (plot_h < 3) plot_h = 3;

        var any = false;
        for (self.series) |s| for (s.items) |v| {
            if (!std.math.isNan(v)) {
                any = true;
                break;
            }
        };

        if (!any) {
            sys.writeAll(sys.stdout_fd, clear_and_home);
            sys.writeAll(sys.stdout_fd, "RTT (ms) over time \xe2\x80\x94 waiting for data\xe2\x80\xa6\n");
            return;
        }

        const slices = self.gpa.alloc([]const f64, self.series.len) catch return;
        defer self.gpa.free(slices);
        for (self.series, 0..) |s, i| slices[i] = s.items;

        const graph = asciigraph.plotMany(self.gpa, slices, self.labels, plot_h, "RTT (ms) over time") catch return;
        defer self.gpa.free(graph);
        sys.writeAll(sys.stdout_fd, clear_and_home);
        sys.writeAll(sys.stdout_fd, graph);
    }

    pub fn final(self: *ChartReporter, ts: rfc3339.Time, rows: []const TargetCycle) void {
        if (self.started) {
            sys.writeAll(sys.stdout_fd, leave_alt_screen ++ show_cursor);
        }
        writeLogLines(self.gpa, ts, rows, true);
    }
};

pub const Reporter = union(enum) {
    log: LogReporter,
    chart: ChartReporter,

    pub fn cycle(self: *Reporter, ts: rfc3339.Time, rows: []const TargetCycle) void {
        switch (self.*) {
            inline else => |*r| r.cycle(ts, rows),
        }
    }
    pub fn final(self: *Reporter, ts: rfc3339.Time, rows: []const TargetCycle) void {
        switch (self.*) {
            inline else => |*r| r.final(ts, rows),
        }
    }
    pub fn deinit(self: *Reporter) void {
        switch (self.*) {
            .chart => |*c2| c2.deinit(),
            .log => {},
        }
    }
};

const TermSize = struct { w: usize, h: usize };

fn termSize() TermSize {
    var ws: posix.winsize = undefined;
    const rc = c.ioctl(sys.stdout_fd, posix.T.IOCGWINSZ, @intFromPtr(&ws));
    if (rc != 0 or ws.col == 0 or ws.row == 0) return .{ .w = 80, .h = 24 };
    return .{ .w = ws.col, .h = ws.row };
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "resolveDisplay" {
    try std.testing.expect(resolveDisplay("log", false) == .ok);
    try std.testing.expect(resolveDisplay("chart", false) == .not_tty);
    try std.testing.expect(resolveDisplay("chart", true) == .ok);
    try std.testing.expect(resolveDisplay("bogus", true) == .invalid);
}
